feat: Technical review — Related website sets docs (#30500)
mdn/content: files/en-us/web/api/document/requeststorageaccessfor/index.md

### Exceptions

- `InvalidStateError` {{domxref("DOMException")}}
  - : Thrown if the current {{domxref("Document")}} is not yet active.
- `NotAllowedError` {{domxref("DOMException")}}
  - : Thrown if:
    - The document's window is not a [secure context](/en-US/docs/Web/Security/Secure_Contexts).
    - The document is not the top-level document.
    - The document has a `null` origin.
    - The supplied `requestedOrigin` is [opaque](https://html.spec.whatwg.org/multipage/browsers.html#concept-origin-opaque).
    - The top-level site and embedded site are not in the same [related website set](/en-US/docs/Web/API/Storage_Access_API/Related_website_sets).
    - The embedding {{htmlelement("iframe")}} is sandboxed, and the `allow-storage-access-by-user-activation` token is not set.
    - Usage is blocked by a {{httpheader("Permissions-Policy/storage-access", "storage-access")}} [Permissions Policy](/en-US/docs/Web/HTTP/Permissions_Policy).
    - Usage is denied by the user agent's permission request to use the API.
- `TypeError`
  - : Thrown if `requestedOrigin` is not a valid URL.

## Description

> `requestStorageAccessFor()` is intended to address challenges in adopting the Storage Access API on top-level sites that use cross-site images or scripts requiring cookies. It can enable third-party cookie access for cross-site resources embedded in {{htmlelement("iframe")}}s, and directly embedded e.g. via {{htmlelement("img")}}s or {{htmlelement("script")}} elements.

Reviewer: ```suggestion
`requestStorageAccessFor()` is intended to address challenges in adopting the Storage Access API on top-level sites that use cross-site images or scripts requiring cookies. It can enable third-party cookie access for directly embedded cross-site resources, e.g. via {{htmlelement("img")}}s or {{htmlelement("script")}} elements.
```

> This method is relevant to user agents that by default block access to [third-party](/en-US/docs/Web/HTTP/Cookies#third-party_cookies), [unpartitioned](/en-US/docs/Web/API/Storage_Access_API#unpartitioned_versus_partitioned_cookies) cookies to improve privacy (e.g. to prevent tracking), and is part of the [Storage Access API](/en-US/docs/Web/API/Storage_Access_API).

Reviewer: ```suggestion
This method is relevant to user agents that by default block access to [third-party](/en-US/docs/Web/HTTP/Cookies#third-party_cookies), [unpartitioned](/en-US/docs/Web/API/Storage_Access_API#unpartitioned_versus_partitioned_cookies) cookies to improve privacy (e.g. to prevent tracking), and is a proposed extension of the [Storage Access API](/en-US/docs/Web/API/Storage_Access_API).
```

I'd call it a proposed extension, rather than saying it's part of the Storage Access API, since it doesn't have cross-browser support (unlike the rest of the Storage Access API, which does have cross-browser support).

---
title: "Document: requestStorageAccessFor() method"
short-title: requestStorageAccessFor()
slug: Web/API/Document/requestStorageAccessFor
page-type: web-api-instance-method
status:
  - experimental
browser-compat: api.Document.requestStorageAccessFor
---

{{APIRef("Storage Access API")}}{{seecompattable}}

> The **`requestStorageAccessFor()`** method of the {{domxref("Document")}} interface allows top-level sites to request third-party cookie access on behalf of embedded content originating from another site in the same [related website set](/en-US/docs/Web/API/Storage_Access_API/Related_website_sets). It Returns a {{jsxref("Promise")}} that resolves if the access was granted, and rejects if access was denied.

Reviewer: ```suggestion
The **`requestStorageAccessFor()`** method of the {{domxref("Document")}} interface allows top-level sites to request third-party cookie access on behalf of embedded content originating from another site in the same [related website set](/en-US/docs/Web/API/Storage_Access_API/Related_website_sets). It returns a {{jsxref("Promise")}} that resolves if the access was granted, and rejects if access was denied.
```

This is intended to address challenges in adopting the Storage Access API on top-level sites that use cross-site images or {{htmlelement("script")}} elements requiring cookies.

`requestStorageAccessFor()` is relevant to user agents that by default block access to [third-party cookies](/en-US/docs/Web/HTTP/Cookies#third-party_cookies), [unpartitioned](/en-US/docs/Web/API/Storage_Access_API#unpartitioned_versus_partitioned_cookies) cookies to improve privacy (e.g. to prevent tracking), and is part of the [Storage Access API](/en-US/docs/Web/API/Storage_Access_API).

> > **Note:** To check whether permission to access third-party cookies has already been granted via `requestStorageAccessFor()`, you can call {{domxref("Permissions.query()")}}, specifying the feature name `"top-level-storage-access"`. Note that this is different to the feature name used for the regular {{domxref("Document.requestStorageAccess()")}} method, which is `"storage-access"`.

Reviewer: Might be good to mention that the call to `navigator.permissions.query` also has to specify the embedded origin, e.g. `navigator.permissions.query({name: 'top-level-storage-access', requestedOrigin: 'https://www.foo.com'})`.

Author: Good call; I've added this to the page.

Author: In my next commit.

## Syntax

```js-nolint

> requestStorageAccessFor(requestedOrigin)

Reviewer: ```suggestion
document.requestStorageAccessFor(requestedOrigin)
```

Author: While I think this is a good suggestion, it isn't the style MDN uses. We always just show the immediate syntax of the method in question, without the object chain required to access it. The examples then show the actual usage.

Reviewer: Ah, I didn't realize that. SGTM, no need to break from pattern then.

```

### Parameters

- `requestedOrigin`
  - : A string representing the URL of the origin you are requesting third-party cookie access for.

### Return value

A {{jsxref("Promise")}} that fulfills with `undefined` if the access to third-party cookies was granted, and rejects if access was denied.

`requestStorageAccessFor()` requests are automatically denied unless the embedded content is currently processing a user gesture such as a tap or click ({{Glossary("transient activation")}}), or unless permission was already granted previously. If permission was not previously granted, they need to be run inside a user gesture-based event handler. The user gesture behavior depends on the state of the promise:

- If the promise resolves (i.e. if permission was granted), then the user gesture has not been consumed, so the script can subsequently call APIs that require a user gesture.

> - If the promise rejects (i.e. permission was not granted), then the user gesture has been consumed, so the script can't do anything that requires a gesture. This is intentional protection against abuse — it prevents scripts from calling `requestStorageAccessFor()` in a loop until the user accepts the prompt.

Reviewer: Strictly speaking, there is no prompt related to `requestStorageAccessFor`, since Chrome doesn't support a prompting experience and other browsers don't support the API.

Author: OK, gotcha. In my next commit, I fixed this by making a small alteration to the last sentence so that it doesn't mention a gesture, but let me know if you think I should go further.

(I also considered deleting both bullets and the sentence "The user gesture behavior depends on the state of the promise:", but I wasn't sure if they were still relevant or not).

Reviewer: The alteration looks good to me. I think it's good to mention the behavior around user gestures, since that's still web-visible behavior.

Author: OK, cool.

### Exceptions

- `InvalidStateError` {{domxref("DOMException")}}
  - : Thrown if the current {{domxref("Document")}} is not yet active.
- `NotAllowedError` {{domxref("DOMException")}}

>   - : Thrown if: - The document's window is not a [secure context](/en-US/docs/Web/Security/Secure_Contexts). - The document or the top-level document has a `null` origin. - The top-level site and embedded site are not in the same [related website set](/en-US/docs/Web/API/Storage_Access_API/Related_website_sets). - The embedding {{htmlelement("iframe")}} is sandboxed, and the `allow-storage-access-by-user-activation` token is not set. - Usage is denied by the user agent's permission request to use the API.

Reviewer: ```suggestion
  - : Thrown if:
    - The document's window is not a [secure context](/en-US/docs/Web/Security/Secure_Contexts).
    - The document is not the top-level document.
    - The document has a `null` origin.
    - The supplied `requestedOrigin` is [opaque](https://html.spec.whatwg.org/multipage/browsers.html#concept-origin-opaque).
    - The top-level site and embedded site are not in the same [related website set](/en-US/docs/Web/API/Storage_Access_API/Related_website_sets).
    - The embedding {{htmlelement("iframe")}} is sandboxed, and the `allow-storage-access-by-user-activation` token is not set.
    - Usage is denied by the user agent's permission request to use the API.
```

>     `TypeError`
>   - : Thrown if `requestedOrigin` is not a valid URL.

Reviewer: ```suggestion
```

Author: Should this definitely be removed? In [the spec](https://privacycg.github.io/requestStorageAccessFor/#the-document-object), the `requestStorageAccessFor()` algorithm includes the lines:

> Let parsedURL be the the result of running the [URL parser](https://url.spec.whatwg.org/#concept-url-parser) on requestedOrigin.
> If parsedURL is failure, [reject](https://webidl.spec.whatwg.org/#reject) p with a "[TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror)" [DOMException](https://webidl.spec.whatwg.org/#idl-DOMException) and return p.

Reviewer: Ah, good call! This looks like a bug in the Chromium code; I've opened https://crbug.com/1505841 to track it.
